Coordinate transaction end markers removal

bashtanov opened the pull request:
Various corner cases require coordination of transaction end markers removal to be separate from one for tombstones.
See [the write-up](https://docs.google.com/document/d/1lL8pjzXhrPCaEim4OhdQoJ5Z9dpk0emKZ-ZPJUAUZgY/edit?tab=t.0) for details.

Currently this PR is based on https://github.com/redpanda-data/redpanda/pull/28530, which is to be changed to use a config property instead of a feature flag.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

WillemKauf committed:
`storage`: reduce calls to feature table and cluster config...
...for checking enablement of tx batch removal.

It seems like there was previously the potential for a race between
enablement of this setting and compaction on the granularity of
a single segment (which would result in undefined behavior if we were
able to e.g. query `false` when attempting to remove a fence batch
but then querying `true` when attempting to remove a control batch),
it wasn't outright obvious from a glance at the code.

Rework this code so that calls to `is_tx_batch_compaction_enabled()` are
minimized in the storage layer- at most once per instanation of
`copy_data_segment_reducer` and `should_keep()` lambdas to be certain that
the above situation cannot happen.

The other locations which call to `is_tx_batch_compaction_enabled()` are
when considering if a `segment` needs a rewrite in sliding window compaction
(if this races with the cluster config being toggled, nothing bad could happen)
as well as calls to `maybe_rebuild_compaction_index()` (once again, no potential
pitfalls if this were to race).

bashtanov committed:
r/types: add MXFO and MXRO to coordinated compaction RPC types

These will serve the same purpose as MCCO/MTRO for tombstone, but for
transaction end markers. For details see the comment added to
compaction_coordinator.cc in the further commit.

bashtanov committed:
r/compaction_coordinator: coordinate offsets for transaction end removal

Originally compaction_coordinator calculated safe offset for tombstone
removal (MTRO) based on cleanly compacted offsets on replicas (MCCOs).

Adde the same logic for removal of transaction end markers (MXRO) based
on max offset free from other transaction-related data (MXFO).

Rename identifiers from specific mcco/mtro to more generic "local
replica offsets" and "group offsets". Unfortunately RPC names cannot be
renamed.

For details see the comment in compaction_coordinator.cc

bashtanov committed:
storage/segment_index: negate may_have_transaction_data_or_fence_batches

For uniformness with `has_clean_compact_timestamp` create a predicate
return true for segments clean from transaction data/fences and false
for those having potential for cleaning.

bashtanov committed:
storage/log: add fn to calcluate max prefix free from tx data and fences

bashtanov committed:
s/types/stm_manager: add max_transaction_end_remove_offset

to pass this value from compaction coordinator (part of raft)
to housekeeping

bashtanov committed:
s/housekeeping_config: add max_tx_end_remove_offset

Also rename it from max_tx_remove_offset to max_tx_end_remove_offset in
compaction config to specify that it only affects commit & abort batches

bashtanov committed:
rptest/tombstone_removal_e2e: handle OfflineLogViewer/segment-merge race

bashtanov committed:
s/log_manager: improve logging in housekeeping_scan

bashtanov committed:
rptest/log_compaction_test: improve logging

log exceptions with stack traces

bashtanov committed:
rptest/log_compaction_test: improve tx removal upgrade tests

One of the scenarios needing may_have_transaction_data_or_fence_batches
segment index flag is the following:
- start on older version
- segment 1 to contain tx fence batch
- segment 2 to contain tx data
- compact them all
- upgrade
- produce commit/abort batches to segment 3.
This proves the need for the flag to clear xact data from segment 2.

Another test added is a "2-segment scenario"
- start on older version
- segment 1 to contain tx fence batch
- segment 2 to contain tx commit batch
- cleanly compact segment 1
- upgrade the cluster
- replicate segment 1 to the new replica: it will contain fence batch
- remove tx batches on the leader
- let segment 2 recover on the replica: it won't contain commit batch.
This proves the need for separate coordination of offsets for tx data.

In the test both scenarios use more segments, as producing or recovering
a specific number of segments is not straightforward, we only can
control these by estimating data sizes and transfer rates.

bashtanov committed:
r/compaction_coordinator: fix initial timer arming

it was not armed before, so it only worked after leadership or config
changed

bashtanov committed:
c/cluster_utils: correct reporting max tombstone remove offset

report actual MTRO instead of max removable log offset

WillemKauf committed:
`config`: deprecate `log_compaction_disable_tx_batch_removal`

And add `log_compaction_tx_batch_removal_enabled`.

Unfortunately this previous cluster config shipped early, and with
the wrong default, so to properly gate the behavior of this property
and enabling backport the feature, we have to deprecate the old config
and add a new one which defaults to off (`false`, in the new case).

WillemKauf committed:
`compaction`: add `is_tx_batch_compaction_enabled()`

To isolate the logic of enabling/disabling compaction of control batches
in one call site.

WillemKauf committed:
`testing`: enable `tx_batch_compaction` in fixture and ducktape tests

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--13bc77d6-033b-4f62-9022-73bc42608447-->
<!--TEST RESULTS-->
<!--de3a6059-befa-4d2c-806d-5e878770a4b0-->
<!--TEST RESULTS-->
<!--ef2b847b-5341-4d69-9dc1-b095cb389951-->
<!--TEST RESULTS-->
<!--ecac79f2-a77f-4ba8-97a7-79017fca0ad5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78189">build#78189</a></summary>


| test_class                 | test_method                   | test_arguments                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                           |
|:---------------------------|:------------------------------|:-----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PrefixTruncateRecoveryTest | test_prefix_truncate_recovery | {"acks": -1, "start_empty": false} | integration | https://buildkite.com/redpanda/redpanda/builds/78189#019b3551-e3c1-4aea-aff4-85c51ef3b64c | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PrefixTruncateRecoveryTest&test_method=test_prefix_truncate_recovery |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78195">build#78195</a></summary>


| test_class                      | test_method                      | test_arguments                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                           |
|:--------------------------------|:---------------------------------|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDLQTest                 | test_dlq_table_for_mixed_records | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/78195#019b3627-7548-4338-91b6-28edc6ab825e | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0017, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_mixed_records         |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test         | {"cloud_storage_type": 1, "mixed_versions": false}                              | integration | https://buildkite.com/redpanda/redpanda/builds/78195#019b3627-7549-46e0-a43b-596290362353 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78260">build#78260</a></summary>


| test_class                 | test_method             | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                  |
|:---------------------------|:------------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ScalingUpTest              | test_fast_node_addition | null               | integration | https://buildkite.com/redpanda/redpanda/builds/78260#019b3e78-8c36-4adb-9555-00885e09dc7f | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0336, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition           |
| ShadowLinkingRandomOpsTest | test_node_operations    | {"failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/78260#019b3e6e-4131-413e-bd76-4b1da6e3ce42 | FLAKY         | 8/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0027, p0=0.0003, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78311">build#78311</a></summary>


| test_class                 | test_method          | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/78311#019b48c9-4e92-45d9-a98c-a926d4bcbefc | FLAKY         | 19/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0116, p0=0.2076, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3917, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<!--/body-->

bashtanov commented on tests/rptest/tests/log_compaction_test.py:
```diff
@@ -907,71 +995,310 @@ def test_tx_control_batch_removal(self):
-class LogCompactionTxRemovalUpgradeTest(LogCompactionTxRemovalTestBase):
-    def __init__(self, test_context):
+class LogCompactionTxRemovalUpgradeTestBase(LogCompactionTxRemovalTestBase):
+    def __init__(self, test_context, initial_version):
         super().__init__(test_context)
```
No, why? Parent's ctor argument has a default of None exactly for the purpose of calling without a value.

bashtanov committed:
c/data-migrate/backend: fix comment typo

bashtanov commented on src/v/storage/types.h:
```diff
@@ -494,7 +499,7 @@ struct housekeeping_config {
       model::offset max_collect_offset,
       model::offset max_tombstone_remove_offset,
-      model::offset max_tx_remove_offset,
+      model::offset max_tx_end_remove_offset,
```
no, that's okay, the line that sets it was removed, but another one was added to the `housekeeping_config` ctor call

bashtanov commented on src/v/raft/follower_states.h:
```diff
@@ -20,8 +20,9 @@ namespace raft {
       : node_id(node)
-      , mcco_getter{std::make_unique<void_executor>()}
-      , mtro_sender{std::make_unique<void_executor>()} {}
+      , coordinated_compaction_offsets_getter{std::make_unique<void_executor>()}
```
I'm not a fan of our 80-char line length limit either, but that's how we have clang-format set up.

bashtanov commented on tests/rptest/tests/log_compaction_test.py:
```diff
@@ -907,71 +996,310 @@ def test_tx_control_batch_removal(self):
                 failed_test_cases.append(e)
         assert len(failed_test_cases) == 0, (
             f"Expected 0 failed test cases, got {len(failed_test_cases)}"
         )
 
 
-class LogCompactionTxRemovalUpgradeTest(LogCompactionTxRemovalTestBase):
-    def __init__(self, test_context):
+class LogCompactionTxRemovalUpgradeTestBase(LogCompactionTxRemovalTestBase):
+    def __init__(self, test_context, initial_version):
         super().__init__(test_context)
-        # Version before `may_have_transactional_batches` was added.
-        self.initial_version: RedpandaVersionLine = (25, 1)
-        # Version before tx removal was added to compaction.
-        self.may_have_tx_batch_version: RedpandaVersionLine = (25, 2)
+        self.initial_version: RedpandaVersion = initial_version
 
     def setUp(self):
+        self.redpanda._installer.start()
         self.redpanda._installer.install(self.redpanda.nodes, self.initial_version)
         self.redpanda.start()
         self.admin = Admin(self.redpanda)
 
-    def upgrade_to_version(self, version):
+    def upgrade_to_version(self, target_version):
         def logical_version():
             return self.admin.get_features()["cluster_version"]
 
-        current_active = logical_version()
-        self.redpanda._installer.install(self.redpanda.nodes, version)
+        def cluster_converged():
+            f = self.admin.get_features()
+            return f["cluster_version"] == f["node_latest_version"]
+
+        self.redpanda._installer.install(self.redpanda.nodes, target_version)
         self.redpanda.restart_nodes(self.redpanda.nodes)
 
         wait_until(
-            lambda: logical_version() > current_active and self.redpanda.healthy(),
+            lambda: cluster_converged() and self.redpanda.healthy(),
```
@oleiman this amends your recent change. I needed this to support minor version upgrades which don't increase logical version. Any concerns?

oleiman replied:
honestly if the test passes it is almost certainly fine (was broken in #28949 before my change). i'll take a closer look though.

WillemKauf commented on src/v/storage/types.h:
```diff
@@ -509,13 +511,13 @@ struct housekeeping_config {
           max_bytes_in_log,
           max_collect_offset,
           max_tombstone_remove_offset,
+          max_tx_end_remove_offset,
```
The blast radius of this commit shows exactly why I wrote this `make_config()` file instead of adding _another_ parameter to the constructors.

I guess it's nice to have it cleaned up like this, but I hope we don't have to go through this process again.

bashtanov replied:
> I hope we don't have to go through this process again

Does it mean you won't mind me keeping settings for retention interval for tombstones separate despite them being always equal now? :wink:

WillemKauf commented on src/v/raft/compaction_coordinator.cc:
```diff
@@ -133,20 +134,32 @@ model::offset compaction_coordinator::get_max_tombstone_remove_offset() const {
 compaction_coordinator::base_interval() const {
-    auto retention = _log->config().tombstone_retention_ms();
+    constexpr auto max = std::chrono::milliseconds::max();
+    auto retention = std::min(
```
thoughts on just using `log->config().delete_retention_ms()` here?

bashtanov replied:
:shrug: makes sense with current implementations of both, but who knows how they change? I perceive `delete_retention_ms()` as conceptually private to `ntp_config`.

BTW I'm not sure I understand why removing tx end markers is safe with tiered storage. But that's probably not for a PR discussion.

WillemKauf commented on src/v/raft/tests/coordinated_compaction_test.cc:
```diff
@@ -88,6 +88,8 @@ struct coco_fixture : raft_fixture {
       model::offset max_collect_offset = model::offset::max()) {
         model::offset max_tombstone_remove_offset
           = n.raft()->log()->stm_manager()->max_tombstone_remove_offset();
+        model::offset max_tx_end_remove_offset
```
Any extra expectations to add to this test around `MXFO`?

bashtanov replied:
I added some, but now we'll need @mmaslankaprv's blessing for exposing `disk_log_impl` in `raft_fixture_base`

WillemKauf reviewed (commented):
Looks good to my eyes, haven't spotted anything out of the ordinary.

oleiman commented on tests/rptest/tests/log_compaction_test.py:
```diff
@@ -919,68 +1003,300 @@ def test_tx_control_batch_removal(self):
 
-        current_active = logical_version()
-        self.redpanda._installer.install(self.redpanda.nodes, version)
+        def cluster_converged():
+            f = self.admin.get_features()
+            return f["cluster_version"] == f["node_latest_version"]
```
yeah seems legit if it works

bharathv commented on src/v/config/configuration.cc:
```diff
@@ -1130,11 +1130,13 @@ configuration::configuration()
+      "log_compaction_tx_batch_removal_enabled",
+      "Enables removal of transactional control batches during compaction. "
+      "These batches are removed according to a topic's configured "
+      "delete.retention.ms.",
```
nit: (this came up today). Perhaps worth mentioning that this only happens for compact/delete,compact topics?

bashtanov committed:
raft/tests/raft_fixture_base: expose disk log

`raft_fixture_base` uses failure-injecting wrapper for the log. For
tinkering with storage layer expose a pointer to the underlying log
object as well.

bashtanov committed:
raft/tests/coordinated_compaction: test tx batch removal coordination

1. check that tx-free and tx end marker removal offsets behave the same
as max cleanly-compacted offset and max tombstone removal offset
respectively.
2. check that if segments are self-compacted only but not cleanly
compacted then transactions will be allowed to remove, unlike tombstones

bashtanov committed:
s/compaction: disable tx end marker removal when tiered storage is on

It's subject to the same race condition as tombstone removal

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/78260-->
<!--44e5df72-5265-4fa4-959e-690c264c42c7-->
<!--title-->
## Retry command for [Build#78260](https://buildkite.com/redpanda/redpanda/builds/78260)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

bashtanov commented:
failure unrelated: https://redpandadata.atlassian.net/browse/CORE-15080

WillemKauf reviewed (dismissed):
Great stuff, thank you!

bharathv reviewed (commented):
lgtm

> It's subject to the same race condition as tombstone removal

for my understanding, can you please expand on this a little bit.

bashtanov commented:
/backport v25.3.x

vbotbuildovich commented:
Failed to create a backport PR to v25.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-29043-v25.3.x-597 remotes/upstream/v25.3.x
git cherry-pick -x 1e6bb21e6f be2ab4e6b7 58e7f627ab f0d16fa806 bbd5c5c21d 340be1c0b5 97721dbd75 2260ab0eda 36803febbb 686c8e1782 e96d9cfd66 147f7f61a4 88bcdfb6b5 dc913a2c5b 1cc9bcafc0 7a402429d0 919b22f95d 84e93286e5 c8d6e2354d 1aeb2423a0
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/20456560754).
